src/core/net/mdns.cpp

```diff
@@ -0,0 +1,6011 @@
+//---------------------------------------------------------------------------------------------------------------------
+// otPlatMdns callbacks
+
+extern "C" void otPlatMdnsHandleReceive(otInstance                  *aInstance,
+                                        otMessage                   *aMessage,
+                                        bool                         aIsUnicast,
+                                        const otPlatMdnsAddressInfo *aAddress)
+{
+    AsCoreType(aInstance).Get<Core>().HandleMessage(AsCoreType(aMessage), aIsUnicast, AsCoreType(aAddress));
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+// Core
+
+const char Core::kLocalDomain[]         = "local.";
+const char Core::kUdpServiceLabel[]     = "_udp";
+const char Core::kTcpServiceLabel[]     = "_tcp";
+const char Core::kSubServiceLabel[]     = "_sub";
+const char Core::kServciesDnssdLabels[] = "_services._dns-sd._udp";
+
+Core::Core(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mIsEnabled(false)
+    , mIsQuestionUnicastAllowed(kDefaultQuAllowed)
+    , mMaxMessageSize(kMaxMessageSize)
+    , mInfraIfIndex(0)
+    , mMultiPacketRxMessages(aInstance)
+    , mNextProbeTxTime(TimerMilli::GetNow() - 1)
+    , mEntryTimer(aInstance)
+    , mEntryTask(aInstance)
+    , mTxMessageHistory(aInstance)
+    , mConflictCallback(nullptr)
+    , mNextQueryTxTime(TimerMilli::GetNow() - 1)
+    , mCacheTimer(aInstance)
+    , mCacheTask(aInstance)
+{
+}
+
+Error Core::SetEnabled(bool aEnable, uint32_t aInfraIfIndex)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(aEnable != mIsEnabled, error = kErrorAlready);
+    SuccessOrExit(error = otPlatMdnsSetListeningEnabled(&GetInstance(), aEnable, aInfraIfIndex));
+
+    mIsEnabled    = aEnable;
+    mInfraIfIndex = aInfraIfIndex;
+
+    if (mIsEnabled)
+    {
+        LogInfo("Enabling on infra-if-index %lu", ToUlong(mInfraIfIndex));
+    }
+    else
+    {
+        LogInfo("Disabling");
+    }
+
+    if (!mIsEnabled)
+    {
+        mHostEntries.Clear();
+        mServiceEntries.Clear();
+        mServiceTypes.Clear();
+        mMultiPacketRxMessages.Clear();
+        mTxMessageHistory.Clear();
+        mEntryTimer.Stop();
+
+        mBrowseCacheList.Clear();
+        mSrvCacheList.Clear();
+        mTxtCacheList.Clear();
+        mIp6AddrCacheList.Clear();
+        mIp4AddrCacheList.Clear();
+        mCacheTimer.Stop();
+    }
+
+exit:
+    return error;
+}
+
+template <typename EntryType, typename ItemInfo>
+Error Core::Register(const ItemInfo &aItemInfo, RequestId aRequestId, RegisterCallback aCallback)
+{
+    Error      error = kErrorNone;
+    EntryType *entry;
+
+    VerifyOrExit(mIsEnabled, error = kErrorInvalidState);
+
+    entry = GetEntryList<EntryType>().FindMatching(aItemInfo);
+
+    if (entry == nullptr)
+    {
+        entry = EntryType::AllocateAndInit(GetInstance(), aItemInfo);
+        OT_ASSERT(entry != nullptr);
```

Comment:
1. I cannot find validation of provided names before memory allocation, e.g. we can provide names like `...local.`
2. It feels that asserting if `AllocateAndInit()` is unable to allocate a memory for entry is a little too harsh. Maybe we should just print an error?

Reply:
Thanks for the review (and noticing the typos :)). All fixed in new push.

Both these are intentional design choices for the current implementation of native mDNS in OT to prioritize code simplicity and reduce overhead.
- Regarding checking the names, the native OT mDNS is primarily intended for use by the OT core stack (advertising proxy, discovery proxies, TREL, etc.) where names are expected to be pre-validated.
- Regarding heap allocation failure: We debated this internally and decided on the current model. Handling heap allocation errors in all situations can become complex, requiring significant code to roll back state and release other resources.  Practically, heap allocation failures are often due to memory leaks, where an assert/reset may be more appropriate than letting the device/BR continue to run with degraded functionality. We can revisit this and possibly add heap allocation error-handling mechanisms if the need becomes pressing.

Reply:
Ad 1. I understand the reasoning. Maybe some kind of @note should be added to the otMdnsRegister* APIs to inform an user about a need of prevalidation?

Ad 2. The problem with this particular assert is that for constrained devices it can be easly triggered by registering a lot of services. It feels like we could just return an Error that could be passed to the other party (e.g. SRP client) instead of resetting the device which could cause a re-registration of all services again and potentially lead to a loop.